lib/pbench/test/unit/agent/test_tool_meister.py

```diff
@@ -0,0 +1,241 @@
+        """Test create tar file"""
+        tm = ToolMeister(
+            pbench_install_dir=None,
+            tmp_dir=None,
+            tar_path="tar_path",
+            sysinfo_dump=None,
+            params=self.tm_params,
+            redis_server=None,
+            logger=None,
+        )
+        tmp_dir = agent_setup["tmp"]
+        cp = tm._create_tar(tmp_dir, pathlib.Path(tar_file))
+        assert cp.returncode == 0
+        assert cp.stdout == b""
+
+    def test_create_tar_ignore_warnings(
+        self, agent_setup, mock_tar_no_warnings, logger_fixture
+    ):
+        """Test creating tar with warning=none option specified"""
+        tm = ToolMeister(
+            pbench_install_dir=None,
+            tmp_dir=None,
+            tar_path="tar_path",
+            sysinfo_dump=None,
+            params=self.tm_params,
+            redis_server=None,
+            logger=logger_fixture,
+        )
+        tmp_dir = agent_setup["tmp"]
+
+        cp = tm._create_tar(tmp_dir, pathlib.Path(tar_file))
+        assert cp.returncode == 0
+        assert cp.stdout == b"No error after --warning=none"
+
+    @pytest.mark.parametrize("logger_fixture", [logging.WARNING], indirect=True)
+    def test_create_tar_failure(
+        self, agent_setup, mock_tar_failure, caplog, logger_fixture
+    ):
+        """Test if we can suppress the errors raised during the tar creation"""
+        tm = ToolMeister(
+            pbench_install_dir=None,
+            tmp_dir=None,
+            tar_path="tar_path",
+            sysinfo_dump=None,
+            params=self.tm_params,
+            redis_server=None,
+            logger=logger_fixture,
+        )
+        tmp_dir = agent_setup["tmp"]
+
+        cp = tm._create_tar(tmp_dir, pathlib.Path(tar_file))
+        assert cp.returncode == 1
+        assert cp.stdout == b"Some error running tar command, empty tar creation failed"
+        assert (
+            f"Tarball creation failed with 1 (stdout 'Some error running tar command, empty tar creation failed') on {tmp_dir}: Re-trying now"
+            in caplog.text
+        )
+
+
+class TestSendDirectory:
+    """Test create_tar in send directory of tool meister"""
+
+    @staticmethod
+    def add_http_mock_response(
+        uri: str, code: HTTPStatus = HTTPStatus.OK, text: str = "succeeded"
+    ):
+        responses.add(responses.PUT, uri, status=code, body=text)
+
+    @staticmethod
+    def fake_tar(returncode: int, stdout: bytes):
+        def f(directory: pathlib.Path, tar_file: pathlib.Path):
+            return subprocess.CompletedProcess(
+                args=[], returncode=returncode, stdout=stdout, stderr=None
+            )
+
+        return f
+
+    @responses.activate
+    def test_tar_create_success(self, agent_setup, monkeypatch, logger_fixture):
+        """This test should pass the tar creation in send directory"""
+
+        def fake_rmtree(directory: pathlib.Path):
+            return True
```

Comment:
In a lot of mocks we've put together along the way, we add a mechanism to track whether the call was made. That is, if the call to `rmtree` is part of the method contract, we can use the mock to verify that. E.g.,

```
    rmtree_called = False

    def fake_rmtree(directory: pathlib.Path):
        rmtree_called = True
        return True
```

So that the main test case can then `assert rmtree_called`. (For extra points, we can pay attention to the `directory` parameter and assert that it was called with the proper path.)

But also ... [the documentation](https://docs.python.org/3/library/shutil.html#shutil.rmtree) doesn't specify a return value, only that exceptions are raised on error. Is the `return True` really valid? If the real `shutil.rmtree` doesn't `return True` then our mock shouldn't either.

Reply:
Yeah I should use `pass` here and also nested functions can only access variables from outer functions, but are not allowed to modify them, we need to use something like `nonlocal` to reference that variable inside the inner function to change its value.
So the example would be
```
rmtree_called = False

def fake_rmtree(directory: pathlib.Path):
    nonlocal rmtree_called
    rmtree_called = True
    return True
```